### PR TITLE
Add asio-grpc

### PR DESCRIPTION
**Pull request by Tradias**
* I've followed [this guide](https://hunter.readthedocs.io/en/latest/creating-new/create/cmake.html)
  step by step carefully. **Yes**

### CODE REVIEW

**Comment by Tradias** [comment]
Is there anything I can or must do about the failing pipelines? It seems that it fails to build gRPC, did I hunterize my project incorrectly (see [CMakeLists.txt](https://github.com/Tradias/asio-grpc/blob/master/CMakeLists.txt#L17-L23) and  [AsioGrpcHunter.cmake](https://github.com/Tradias/asio-grpc/blob/master/cmake/AsioGrpcHunter.cmake))? I created a small test project locally that consumes asio-grpc through hunter it everything works successfully.

Also, I know that `hunter_add_version` currently points at the master branch of asio-grpc. I will release a new version (1.6.0) once you can confirm that asio-grpc has been hunterized correctly.

**Comment by NeroBurner** [comment]
I just looked at a failing build, not on the PR itself or the hunterized project. But here are my findings and suspicions

https://github.com/cpp-pm/hunter/runs/6147671151?check_suite_focus=true#step:7:5713

```
/usr/bin/ld: /home/runner/work/hunter/hunter/_testing/Hunter/_Base/eca1947/9dcf50b/8c25157/Install/lib/libcares.a(ares_strerror.c.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
```



you seem to be creating a shared library (is this intended?), but one of the dependencies for the shared library was built as static library without relocation support. The static library needs to be compiled with `-fPIC` flag. This flag is default set in MSVC, but unfortunately not on GCC/Clang (I think because of historic reasons, but I'm not sure). Two ways to fix this

- the dependency (or all dependencies) must set the `-fPIC` flag to compile static libraries (or generally libraries, but -fPIC is redundant for shared libraries)
- use a cmake toolchain file, where `-fPIC` is set as a compiler flag

I prefer the first fix, as the second one is more of a workaround in my honest opinion

**Comment by Tradias** [comment]
Hi, thank you for the quick feedback. Asio-grpc is a header only library and does not need any dependencies to be installed, but in order to conform to the hunter guidelines it should list its runtime dependencies through `hunter_add_package`. At least, that is how I understood it from other PR's (e.g. restinio). One of these dependencies is gRPC, which builds c-ares internally . I think my PR and asio-grpc does not and should not influence how gRPC is being build by hunter. I suspect that the CI for gRPC itself here on hunter would also fail with the same error.

**Comment by NeroBurner** [comment]
ah, the build table https://cpp-pm.github.io/hunter/ (search for gRPC) states, that it builds only with a `fpic` enabled toolchain

This is good, because one toolchain here `gcc-9-cxx17-fpic` successfully builds. `gRPC` additionally was able to build on `gcc-8-cxx17-fpic`. Is gcc-8 not supported by `asio-grpc`? Not a blocker if it isn't

File: docs/packages/pkg/asio-grpc.rst [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  single: networking ; asio-grpc
+
+.. _pkg.asio-grpc:
+
+asio-grpc
+=========
+
+-  `Official <https://github.com/Tradias/asio-grpc>`__
+-  `Example <https://github.com/cpp-pm/hunter/blob/master/examples/asio-grpc/CMakeLists.txt>`__
+-  Added by `Tradias <https://github.com/Tradias>`__ (`pr-N <https://github.com/cpp-pm/hunter/pull/N>`__)
```
Comment (NeroBurner) [inline docs/packages/pkg/asio-grpc.rst:16]:
```suggestion
-  Added by `Tradias <https://github.com/Tradias>`__ (`pr-554 <https://github.com/cpp-pm/hunter/pull/554>`__)
```

File: cmake/projects/asio-grpc/hunter.cmake [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2016-2020, Rahul Sheth, Ruslan Baratov
```
Comment (NeroBurner) [inline cmake/projects/asio-grpc/hunter.cmake:1]: You can add yourself here and update the year

```suggestion
# Copyright (c) 2022, Tradias
```

File: cmake/projects/asio-grpc/hunter.cmake [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+include(hunter_download)
+include(hunter_pick_scheme)
+
+hunter_add_version(
+    PACKAGE_NAME
+    asio-grpc
+    VERSION
+    1.5.1
+    URL
+    "https://github.com/Tradias/asio-grpc/archive/refs/heads/master.zip "
```
Comment (NeroBurner) [inline cmake/projects/asio-grpc/hunter.cmake:17]: just leaving a note here to not forget about the release tag later

File: examples/asio-grpc/boo.cpp [hunk 4]
```diff
@@ -0,0 +1,4 @@
+#include <agrpc/asioGrpc.hpp>
+
+int main() {
+}
```
Comment (NeroBurner) [inline examples/asio-grpc/boo.cpp:4]: could you use something from the header here, to make it double sure, that the include did work?

**Review by NeroBurner — CHANGES_REQUESTED** [review]

**Comment by Tradias** [comment]
Good catch. Asio-grpc requires the `ASIO_GRPC_USE_BOOST_CONTAINER` CMake arg to compile with GCC 8. I could change its default value for compilers known to need it, will do that.

File: examples/asio-grpc/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+project(download-asio-grpc)
+
+# DOCUMENTATION_START {
+hunter_add_package(asio-grpc)
+find_package(asio-grpc CONFIG REQUIRED)
+
+add_executable(boo boo.cpp)
+target_link_libraries(boo PUBLIC asio-grpc::asio-grpc)
+# DOCUMENTATION_END }
```
Comment (NeroBurner) [inline examples/asio-grpc/CMakeLists.txt:18]: In other examples we additionally do a second `find_package()` to make sure the imported library target creation is well behaved. And please add a test if the license of the package is well defined

See https://github.com/cpp-pm/hunter/blob/master/examples/Eigen/CMakeLists.txt#L20 for pointers how to do that

**Review by NeroBurner — CHANGES_REQUESTED** [review]
Nice progress on this PR. Just some small Licensing and testing addition, then we're good to go

File: examples/asio-grpc/CMakeLists.txt (hunk 5 above)
Comment (NeroBurner) [reply examples/asio-grpc/CMakeLists.txt:18]: See https://hunter.readthedocs.io/en/latest/user-guides/cmake-user/using-license-files.html?highlight=HUNTER_INSTALL_LICENSE_FILES#using-license-files for documentation about license files and how hunter finds them

File: examples/asio-grpc/CMakeLists.txt [hunk 6]
```diff
@@ -16,3 +16,20 @@ find_package(asio-grpc CONFIG REQUIRED)
...
+  message(FATAL_ERROR "Licenses not found")
+endif()
+
+message("asio-grpc licenses:")
+foreach(x ${asio-grpc_LICENSES})
+  message("* ${x}")
+  if(NOT EXISTS "${x}")
+    message(FATAL_ERROR "File not found")
+  endif()
+endforeach()
```
Comment (NeroBurner) [inline examples/asio-grpc/CMakeLists.txt:35]: small nitpick, non-blocking: please add a newline at the end of files

```suggestion
endforeach()

```

**Review by NeroBurner — APPROVED** [review]
LGTM thank you for your contribution!

**Pull request merged**